Always display buffer list in tabline

chrisbra opened the pull request:
regardless of how many tabs are open.
fixes #639

Simply set `:let g:airline#extensions#tabline#show_buffers=2`

bling commented:
this is one way to resolve the issue.  another is to implement one of the ideas in the thread which is to have something like:

    split 1 > split 2 > split 3                            tab 2 < tab 1

i implemented a spike a long time ago but never got a chance to finish it.  i just rebased it and pushed to the [639-spike](https://github.com/vim-airline/vim-airline/tree/639-spike) branch.  i got it to the point where it does what is illustrated above, but the separators get drawn incorrectly.  perhaps you can finish it, or take it as inspiration to implement it in a better way.

either way, i suspect users may want both the method you proposed in this PR and the "merged" split/tab approach.

chrisbra commented:
Thanks, I'll look into it soon.

chrisbra commented:
That is actually pretty cool. I am impressed. 
I think I found the problem with the separators. The problem is, first we'll create a hi group
airline_tab_to_airline_tabsel for the left side, than later when creating the right side we do recreate 
an airline_tab_to_airline_tabsel for the right side. But now the colors are wrong for the left side.

I have took your approach and pushed it to the spike branch of chrisbra/vim-airline.

However, I have a questions. It looks like you first wanted to create a combined tabline format, but later decided to use that as the default format for the tabline, when tabs exist. So should we simply make this the new default behaviour?

chrisbra commented:
I had to adjust the tests. It would be nice, if this could be tested, before this will be merged

bling commented:
> However, I have a questions. It looks like you first wanted to create a combined tabline format, but later decided to use that as the default format for the tabline, when tabs exist. So should we simply make this the new default behaviour?

yep, that's exactly what happened.  i started down the path of making it configurable, and once i completed a first pass implementation i figured it made little sense to keep any of the old options around because the new way was so much better.

sappo commented:
The ctrlspace tabline integration already implements this feature. I also had to resolve the separator issue for which I introduced new highlight classes in the `tabline.vim`.

chrisbra commented:
Okay, I think I have reworked @bling  work on branch spike and commited some fixes and squashed into a single commit. I think this is ready to be merged.

thirtythreeforty commented:
@chrisbra I think it would be fantastic if this were merged; great work!  One quick tip from the peanut gallery: make sure you document the new flag `g:airline#extensions#tabline#combined`.

chrisbra commented:
@thirtythreeforty `g:airline#extensions#tabline#combined` is a leftover. I'll remove it.

bling commented:
@chrisbra looks really good!  just one small change and it's good to go.  when you have multiple tabs open it shows `tab 1 <tab 2 <tab 3<` instead of ` tab 1 < tab 2 < tab 3`.

@sappo as a general rule i prefer to have other plugins take care of the heavy lifting.  it's easier on maintenance for all parties involved, and it keeps the goals of airline simpler -- e.g. it's just the target of statusline information.  however, the tabline's been in airline for over 2.5 years now (before ctrlspace gained popularity), so we need to continue supporting the existing user base.

chrisbra commented:
> @chrisbra looks really good! just one small change and it's good to go. when you have multiple tabs open it shows tab 1 <tab 2 <tab 3< instead of tab 1 < tab 2 < tab 3.

That is just because of the close button, isn't it?
E.g. with the default let g:airline#extensions#tabline#show_close_button = 1 it looks like this:
![Imgur](http://i.imgur.com/738McxQ.png)

And setting it to zero it looks like this:
![Imgur1](http://i.imgur.com/oS2r92W.png)

bling commented:
nope, i was referring to `tab 1` having a space after it, but not before, so the text is not centered for each section.

chrisbra commented:
Okay, I added the space, because it looked too close to the separators otherwise. Will change. Thanks.

chrisbra commented:
Okay, updated according to comments. Should apply now cleanly

bling commented:
sorry for the back and forth.  every section should have 1 space on both the left and right side of the word.  e.g.

```
| buffer 1 > buffer 2 > buffer 3               tab 1 < tab 2 < tab 3 |
```

this way the spacing is the same on both the left side (for buffers), and the right (for tabs).  other than that, i'm ready to hit the big green button!!

cHoco commented:
Just tested it and I have a few comments:
1. Seems to break compatibility with Taboo
2. Buffers are only displayed when they are in splits of the current tab, why doesn't it show all the buffers?
3. It'd be nice if the right section preserved the same color as before, or at least the right section (tabs) used different colors than buffers.

chrisbra commented:
Am 2016-02-11 14:35, schrieb Enrico Ghirardi:
> Just tested it and I have a few comments:
> 1. Seems to break compatibility with Taboo

What is that and what does it do?

> 2. Buffers are only displayed when they are in splits of the current
> tab, why doesn't it show all the buffers?

Sounded more useful, to show only the buffers open per tabpage
Thats what tabpagebuflist() returns.

> 3. It'd be nice if the right section preserved the same color as
> before, or at least the right section (tabs) used different colors
> than buffers.

We can fix that.

cHoco commented:
> What is that and what does it do?

https://github.com/gcmt/taboo.vim A plugin that provides ability to rename tabs and styling the name with some formatting. vim-airline integrates with it but it doesn't work with this change.

> Sounded more useful, to show only the buffers open per tabpage
> Thats what tabpagebuflist() returns.

I'm not sure what I'd prefer, because on one hand it makes sense to only show tabpagebuflist() but on the other I know what buffers the tab is presenting because I can already see them. Buffers tab is useful, for me, when I don't have a lot of buffers and I can just move fast between them using `:bn :bp`. At least having an option to decide the behaviour would be nice.


Finally, is there a reason for not keeping around the hold behaviour? When I work on a big project and I need to jump between 7-8 tab pages I would prefer to just show them and not the buffers.

bling commented:
there are many factors at play here.  let's break it down.

1.  showing all tabs and all buffers.  i don't think this makes sense because there isn't enough room for everything. 
2.  showing tabs only.  this is the existing behavior.
3.  showing all tabs, with the panes per tab.  this is this PR.
4.  showing buffers only, regardless of which tab.  this is a configurable setting that's already available.  it could be improved a bit by showing the selected tab, but not more than that because of space restrictions.

there's probably more use cases, but really what i'm trying to get at is that there's not enough room for everything, so we need to come up with the best bang for buck.  unless neovim gives us two tablines to work with, we only have one.

chrisbra commented:
> https://github.com/gcmt/taboo.vim A plugin that provides ability to rename tabs and styling the name with some formatting. vim-airline integrates with it but it doesn't work with this change.

I am looking into it.

> I'm not sure what I'd prefer, because on one hand it makes sense to only show tabpagebuflist() but on the other I know what buffers the tab is presenting because I can already see them. Buffers tab is useful, for me, when I don't have a lot of buffers and I can just move fast between them using :bn :bp. At least having an option to decide the behaviour would be nice.

I'd like to keep it as it, because otherwise we will be running out of space soon. This might even happen with the current implementation, if you have many splits open.

> Finally, is there a reason for not keeping around the hold behaviour? When I work on a big project and I need to jump between 7-8 tab pages I would prefer to just show them and not the buffers.

For me, it was like @bling experience. It just did not make much sense, to keep the old behaviour.

chrisbra commented:
Regarding the taboo plugin not working, I am wondering how it is working for you. Because for me, it will always replace the default 'tabline' setting, because it uses a VimEnter autocommand. We can fix that in airline, but that is another problem, should be done by another PR.

chrisbra commented:
updated according to comments
(different color for inactive tabs, tabs are spaced equally)

cHoco commented:
I think that when there isn't enough space the buffers sections should get truncated, instead right now the tabs sections gets truncated. 

@chrisbra 

> Regarding the taboo plugin not working, I am wondering how it is working for you. Because for me, it will always replace the default 'tabline' setting, because it uses a VimEnter autocommand.

Did you set `let g:taboo_tabline` to `0`? So it doesn't set the tabline during VimEnter.
It broke compatibility because you set name manually instead of using `airline#extensions#tabline#title` as before.

Any reason why you removed the old styling options for tab names? I personally liked them, I think we should keep them.
Proposed change:
``` patch
diff --git a/autoload/airline/extensions/tabline/tabs.vim b/autoload/airline/extensions/tabline/tabs.vim
index e855930..3af5476 100644
--- a/autoload/airline/extensions/tabline/tabs.vim
+++ b/autoload/airline/extensions/tabline/tabs.vim
@@ -2,6 +2,8 @@
 " vim: et ts=2 sts=2 sw=2

 let s:show_close_button = get(g:, 'airline#extensions#tabline#show_close_button', 1)
+let s:show_tab_nr = get(g:, 'airline#extensions#tabline#show_tab_nr', 1)
+let s:tab_nr_type = get(g:, 'airline#extensions#tabline#tab_nr_type', 0)
 let s:close_symbol = get(g:, 'airline#extensions#tabline#close_symbol', 'X')
 let s:spc = g:airline_symbols.space

@@ -63,7 +65,17 @@ function! airline#extensions#tabline#tabs#get()
     else
       let group = 'airline_tab_right'
     endif
-    call b.add_section_spaced(group, '%'.i.'Ttab '.i)
+    let val = '%('
+    if s:show_tab_nr
+      if s:tab_nr_type == 0
+        let val .= (g:airline_symbols.space).'%{len(tabpagebuflist('.i.'))}'
+      elseif s:tab_nr_type == 1
+        let val .= (g:airline_symbols.space).i
+      else "== 2
+        let val .= (g:airline_symbols.space).i.'.%{len(tabpagebuflist('.i.'))}'
+      endif
+    endif
+    call b.add_section(group, val.'%'.i.'T %{airline#extensions#tabline#title('.i.')} %)')
   endfor
   call b.add_raw('%<')
```
restores old tab titles styling and taboo compatibility.
Personally I would also change the `airline#extensions#tabline#title` function to return the manual string `Tab n` instead of the buffer name. 

@bling 
> showing all tabs and all buffers. i don't think this makes sense because there isn't enough room for everything.

I still think it would be useful, I think an option to decide what buffers to show wouldn't hurt. The space occupied by two tabs is almost the same of the old `tabs` label.

chrisbra commented:
Alright @cHoco restored the old behaviour from your comment.

chrisbra commented:
I think it is ready to be merged now.

bling commented:
hmm, if i have taboo installed the tabline extension doesn't get activated on startup.  it would appear that the guard at the top of `tabline.vim` is not enough because taboo sets up a VimEnter autocmd which will overwrite ours...

we'll need something like this:

```
diff --git a/autoload/airline/extensions/tabline.vim b/autoload/airline/extensions/tabline.vim
index de6d55e..2cb06f3 100644
--- a/autoload/airline/extensions/tabline.vim
+++ b/autoload/airline/extensions/tabline.vim
@@ -9,6 +9,7 @@ let s:ignore_bufadd_pat = get(g:, 'airline#extensions#tabline#ignore_bufadd_pat'
 let s:taboo = get(g:, 'airline#extensions#taboo#enabled', 1) && get(g:, 'loaded_taboo', 0)
 if s:taboo
   let g:taboo_tabline = 0
+  autocmd VimEnter * call s:toggle_on()
 endif
 
 let s:ctrlspace = get(g:, 'CtrlSpaceLoaded', 0)
```

one concern i have is that the buffer and tab colors are the same, and without labels on either side it might be confusing to users.  but we can't predict every use case, so after the above change (unless you think there's a better way), we can hit the green button and see what the masses think :smile:

chrisbra commented:
@bling I opened issue gcmt/taboo.vim#12 about it. I personally do not like adding another autocommand, but would rather use:
```patch
diff --git a/autoload/airline/extensions/tabline.vim b/autoload/airline/extensions/tabline.vim
index 0447952..245a8ed 100644
--- a/autoload/airline/extensions/tabline.vim
+++ b/autoload/airline/extensions/tabline.vim
@@ -8,6 +8,7 @@ let s:ignore_bufadd_pat = get(g:, 'airline#extensions#tabline#ignore_bufadd_pat'
 let s:taboo = get(g:, 'airline#extensions#taboo#enabled', 1) && get(g:, 'loaded_taboo', 0)
 if s:taboo
   let g:taboo_tabline = 0
+  au! taboo VimEnter
 endif

 let s:ctrlspace = get(g:, 'CtrlSpaceLoaded', 0)
```

But in either case, I'll leave it to you to push that button ;)

bling commented:
gonna merge this and see how people respond!

zdm commented:
Hey, guys.

I prefer to work with tabs, not buffers. What I should write into vimrc to restore old behavior, when only tabs was shown in tabline?

zhuochun commented:
Same question on how to restore old behavior with show tabs only.

bling commented:
@zdm @zhuochun please continue discussion at #1043.

r-tanner-f commented:
> gonna merge this and see how people respond!

Just updated for this. +1 Love it! Thank you.
